turnstile-example/turnstile/examples/cs478/current/stlc.rkt

```diff
@@ -442,3 +446,23 @@
   #;  [⊢ (if- (inl-? e-)
           (let- ([x- (inl--val e-)]) e1-)
           (let- ([y- (inr--val e-)]) e2-)) ⇒ τ])
+
+(provide typed-match)
+(define-typerule typed-match
+  [(_ (x:id e) b) ≫
+   [⊢ e ≫ e- ⇒ τ]
+   [[x ≫ x- : τ] ⊢ b ≫ b- ⇒ τ-out]
+   ---------------------------------
+   [⊢ (let- ([x- e-]) b-) ⇒ τ-out]]
+  [(_ ([x:id ...] e) b) ≫
+   [⊢ e ≫ e- ⇒ (~× τ-tup ...)]
+   #:fail-unless (= (stx-length #'(x ...)) (stx-length #'(τ-tup ...)))
+   (format "~a pattern variables does not match ~a values in tuple" (stx-length #'(x ...)) (stx-length #'(τ-tup ...)))
+   [[x ≫ x- : τ-tup] ... ⊢ b ≫ b- ⇒ τ-out]
+   -----------------------------------------
+   [⊢ (match- e- [(list- x- ...) b-]) ⇒ τ-out]]
+  [(_ ([x:id ...] e) b) ≫
+   [⊢ e ≫ e- ⇒ (~and τ-rec:type (~parse (~Rec [l = τ-label] ...) #'τ-rec.norm))]
```

Comment:
My understanding is that `τ-rec` is just a pattern variable, and when used on its own has no meaning, whereas `'τ-rec` means the runtime representation. What's the difference between:
```
#'τ-rec
#'τ-rec.norm
```

Reply:
The `norm` is the "normalized" representation of the type, ie it's the expanded type.  It's more useful once we start to have reduction rules in the types.

The `type` stx class is typically used for type positions in the surface syntax, eg a lambda annotation. 
Internally, like here, it's not really needed because you already know `e`'s type is a type. You wouldnt need it here unless you allow something else in your type system